src/conn_sock.c

```diff
@@ -349,27 +349,32 @@ char *socket_parent_dir(gboolean use_full_attach_path, size_t desired_len)
 	return base_path;
 }
 
-
 void schedule_main_stdin_write()
 {
 	schedule_local_sock_write(&local_mainfd_stdin);
 }
 
-void write_back_to_remote_consoles(char *buf, int len)
+static void write_remote_sock(gpointer data, G_GNUC_UNUSED gpointer user_data)
 {
-	if (local_mainfd_stdin.readers == NULL)
+	struct remote_sock_s *remote_sock = (struct remote_sock_s *)data;
+	if (!remote_sock->writable)
 		return;
-
-	for (int i = local_mainfd_stdin.readers->len; i > 0; i--) {
-		struct remote_sock_s *remote_sock = g_ptr_array_index(local_mainfd_stdin.readers, i - 1);
-
-		if (remote_sock->writable && write_all(remote_sock->fd, buf, len) < 0) {
-			nwarn("Failed to write to remote console socket");
-			remote_sock_shutdown(remote_sock, SHUT_WR);
-		}
+	writev_iov_t *iov = (writev_iov_t *)user_data;
+	if (writev_all(remote_sock->fd, iov) < 0) {
+		nwarn("Failed to write to remote console socket");
+		remote_sock_shutdown(remote_sock, SHUT_WR);
 	}
 }
 
+void write_back_to_remote_consoles(stdpipe_t pipe, char *buf, size_t buflen)
+{
+	if (local_mainfd_stdin.readers == NULL)
+		return;
+	struct iovec iov[2] = {{&pipe, 1}, {buf, buflen}};
+	writev_iov_t wviov = {2, 2, iov};
+	g_ptr_array_foreach(local_mainfd_stdin.readers, write_remote_sock, &wviov);
```

Comment:
I think another issue here is that change of order of array iteration.

Previously, the `local_mainfd_stdin.readers` was iterated in the reverse order. I believe the reason was that the `remote_sock_shutdown` can remove the reader from the array (It calls `g_ptr_array_remove`).

The proposed code uses `g_ptr_array_foreach`. The documentation for this says `func must not add elements to or remove elements from the array.`, but this is what can actually happen in the `remote_sock_shutdown`.

I think the way forward is to keep the reverse iteration.